### PR TITLE
Future-proof MIR for dedicated debuginfo.

### PR DESCRIPTION
**Pull request by eddyb**
This is #56231 without the last commit (the one that actually moves to `VarDebuginfo`).
Nothing should be broken, but it should no longer depend on debuginfo for anything else.

r? @nikomatsakis

### CODE REVIEW

File: src/librustc_mir/borrow_check/mod.rs [hunk 1]
```diff
@@ -141,6 +142,40 @@ fn do_mir_borrowck<'a, 'gcx, 'tcx>(
         .as_local_node_id(def_id)
         .expect("do_mir_borrowck: non-local DefId");
 
+    // Gather the upvars of a closure, if any.
+    let upvars: Vec<_> = tcx.with_freevars(id, |freevars| {
```
Comment (nikomatsakis) [inline src/librustc_mir/borrow_check/mod.rs]: This really feels like information that could be encoded in the MIR, no? Well, I guess I can see an argument both ways.

Comment (nikomatsakis) [reply src/librustc_mir/borrow_check/mod.rs]: At minimum I would prefer if we can factor it out from *borrow check* into some kind of helper table that would be more re-usable.

Comment (nikomatsakis) [reply src/librustc_mir/borrow_check/mod.rs]: But I don't know that I want to block the PR on it. In general I feel like the borrow check code is kind of noodly and could use some extractions of this kind.

File: src/librustc/mir/mod.rs [hunk 2]
```diff
@@ -139,16 +139,16 @@ pub struct Mir<'tcx> {
...
-
     /// Mark an argument local (which must be a tuple) as getting passed as
     /// its individual components at the LLVM level.
     ///
     /// This is used for the "rust-call" ABI.
     pub spread_arg: Option<Local>,
 
+    /// Names and capture modes of all the closure upvars, assuming
+    /// the first argument is either the closure or a reference to it.
+    pub upvar_decls: Vec<UpvarDecl>,
```
Comment (nikomatsakis) [inline src/librustc/mir/mod.rs]: btw, I think that if there are fields that are "debuginfo" that you feel other people should not rely on, we should comment that =)

(Like this one, judging from previous commit)

File: src/librustc_mir/interpret/validity.rs [hunk 3]
```diff
@@ -175,13 +176,26 @@ impl<'rt, 'a, 'mir, 'tcx, M: Machine<'a, 'mir, 'tcx>> ValidityVisitor<'rt, 'a, '
...
             ty::Closure(def_id, _) | ty::Generator(def_id, _, _) => {
-                if let Some(upvar) = self.ecx.tcx.optimized_mir(def_id).upvar_decls.get(field) {
-                    PathElem::ClosureVar(upvar.debug_name)
-                } else {
-                    // Sometimes the index is beyond the number of freevars (seen
-                    // for a generator).
-                    PathElem::ClosureVar(Symbol::intern(&field.to_string()))
+                let mut name = None;
+                if let Some(node_id) = self.ecx.tcx.hir.as_local_node_id(def_id) {
+                     self.ecx.tcx.with_freevars(node_id, |freevars| {
```
Comment (nikomatsakis) [inline src/librustc_mir/interpret/validity.rs]: I wonder if this has any overlap with the borrow check code I was noting earlier? (i.e., if some common helper could make both of these more concise)

File: src/librustc/mir/mod.rs [hunk 4]
```diff
@@ -147,7 +147,11 @@ pub struct Mir<'tcx> {
 
     /// Names and capture modes of all the closure upvars, assuming
     /// the first argument is either the closure or a reference to it.
-    pub upvar_decls: Vec<UpvarDecl>,
+    // NOTE(eddyb) This is *strictly* a temporary hack for codegen
+    // debuginfo generation, and will be removed at some point.
+    // Do **NOT** use it for anything else, upvar information should not be
+    // in the MIR, please rely on local crate HIR or other side-channels.
+    pub __upvar_debuginfo_codegen_only_do_not_use: Vec<UpvarDebuginfo>,
```
Comment (nikomatsakis) [inline src/librustc/mir/mod.rs]: lol ok I guess that addresses my earlier concern

Comment (nikomatsakis) [reply src/librustc/mir/mod.rs]: I would like this to be part of the doc comment please though, why use only `//`?

**Review by nikomatsakis — APPROVED** [review]
these changes look fine, though I'd be happier if we could make some nice queries for getting info about the Nth upvar of a closure, so that the code would be a bit cleaner

File: src/librustc_mir/borrow_check/mod.rs [hunk 5]
```diff
@@ -2208,3 +2277,15 @@ impl ContextKind {
         }
     }
 }
+
+#[derive(Debug)]
+crate struct Upvar {
```
Comment (nikomatsakis) [inline src/librustc_mir/borrow_check/mod.rs]: Yeah I think I would prefer to see `closure_upvars(def_id, n) -> Upvar` that returns this information "nicely collected"....  (or `closure_upvars(def_id) -> &[Upvar]`). 

But I won't hold up the PR on it. 

r=me, depending on how you feel @eddyb

File: src/librustc/mir/mod.rs (hunk 4 above)
Comment (eddyb) [reply src/librustc/mir/mod.rs]: I got used to only using `//` for NOTE/FIXME/HACK, but I'll go change it.

**Comment by Aaronepower** [comment]
Triage; @eddyb  Hello, have you been able to get back to this PR?

**Comment by Aaronepower** [comment]
Triage; @eddyb Hello again, have you been able to get back to this PR?

**Comment by TimNN** [comment]
Ping from triage @eddyb: What is the status of this PR?

**Comment by Dylan-DPC** [comment]
ping from triage @eddyb Unfortunately we haven't heard from you on this in a while, so I'm closing the PR to keep things tidy. Don't worry though, if you'll have time again in the future please reopen this PR, we'll be happy to review it again!

**Comment by eddyb** [comment]
I've started rebasing this, I'll try to get it merged next week.

**Comment by eddyb** [comment]
I think I want to get rid of `rustc_mir::borrowck::Upvar`, I might be able to reuse some of the new information that's in the `TypeckTables` now.

**Comment by Dylan-DPC** [comment]
ping from triage @eddyb any updates? you have conflicts to resolve

**Comment by eddyb** [comment]
Regarding https://github.com/rust-lang/rust/pull/56278#issuecomment-464027759, the `TypeckTables` field in question is from issue #56905 / PR #56906, i.e.:
https://github.com/rust-lang/rust/blob/e4e032a0ae82d7db4f99872ff98626af2941c4a5/src/librustc_mir/build/mod.rs#L628-L640

File: src/librustc_mir/borrow_check/nll/explain_borrow/mod.rs [hunk 6]
```diff
@@ -273,11 +273,7 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
                 if mir.local_decls[local].name.is_some() {
                     if let Some((WriteKind::StorageDeadOrDrop, place)) = kind_place {
                         if let Place::Base(PlaceBase::Local(borrowed_local)) = place {
-                            let dropped_local_scope = mir.local_decls[local].visibility_scope;
-                            let borrowed_local_scope =
-                                mir.local_decls[*borrowed_local].visibility_scope;
-
-                            if mir.is_sub_scope(borrowed_local_scope, dropped_local_scope)
+                             if mir.local_decls[*borrowed_local].name.is_some()
```
Comment (matthewjasper) [inline src/librustc_mir/borrow_check/nll/explain_borrow/mod.rs]: Why isn't this checking `is_user_variable`?

Comment (eddyb) [reply src/librustc_mir/borrow_check/nll/explain_borrow/mod.rs]: Oh, looks like `is_user_variable` is newer than this PR originally, and is actually part of the set of things I meant to review at some point.

**Pull request merged**